File: Jens/day04.clj
(def input (-> "inputs/day04.txt"
               slurp 
               (clojure.string/split #"\n\n")))

(defn valid? [passport]
  (= 7 (count (re-seq #"(?:byr|iyr|eyr|hgt|hcl|ecl|pid)\:" passport))))

(defn day4-1 []
  (count (filter valid? input)))

(defn valid-field? [field]
  (let [check-year (fn [x l u] (and (= 4 (count x))
                                    (<= l (Integer/parseInt x) u)))
        val (nth field 2)]
    (condp = (second field)
      "byr" (check-year val 1920 2002)
      "iyr" (check-year val 2010 2020)
      "eyr" (check-year val 2020 2030)
      "hgt" (let [[_ height unit] (re-find #"(\d+)(cm|in)" val)]
              (condp = unit
                "cm" (<= 150 (Integer/parseInt height) 193)
                "in" (<= 59 (Integer/parseInt height) 76)
                false))
      "hcl" (and (re-find #"#[\d|[a-f]]{6}" val)
                 (= 7 (count val)))
      "ecl" (re-find #"(?:amb|blu|brn|gry|grn|hzl|oth)" val)
      "pid" (and (re-find #"\d{9}" val)
                 (= 9 (count val))))))

(defn valid2? [passport]
  (let [fields (re-seq #"(byr|iyr|eyr|hgt|hcl|ecl|pid):(#{0,1}[\d|a-z]+)" passport)]
    (and (= 7 (count fields))
<<<<<<< HEAD
<<<<<<< HEAD
         (every? valid-field? fields))))

(defn day4-2 []
=======
         (every? #(and (some? %) (not (false? %))) (map valid-field? fields)))))

(defn day4-1 []
>>>>>>> 2749eb554812233338a5fe55b73d80ed03ae2ac9
=======
         (every? #(and (some? %) (not (false? %))) (map valid-field? fields)))))

(defn day4-1 []
>>>>>>> 2749eb554812233338a5fe55b73d80ed03ae2ac9
  (count (filter valid2? input)))

File: Erik/day04.py
import re
password_fields = ["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid", "cid"]
check_fields = ["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"]

with open("Advent-of-Code-2020/Erik/inputs/day04.txt", "r") as f:
#with open("Advent-of-Code-2020/Erik/inputs/day04-t1.txt", "r") as f:
    raw_data2 = f.read().split("\n\n")

def make_dict(raw):
    dict_list = []
    listed_data = [x.split() for x in raw]
    for passport in listed_data:
        temp_dict = dict()
        for line in passport:
            key, val = line.split(":", 1)
            temp_dict.update({key:val})
        dict_list.append(temp_dict)
    return dict_list

def has_all_fields(passport_dict, req_fields):
    for field in req_fields:
        if field not in passport_dict:
            return 0
    return 1

def part_one(data, req_fields):
    counter = 0
    for passport in data:
        counter += has_all_fields(passport, req_fields)
    return counter

def validate_passport(passport, check_fields):
    if not has_all_fields(passport, check_fields):
        return False
    for key in passport:
        if key == "byr":
            year = int(passport[key])
            if((year < 1920) or (year > 2002)):
                return False
        elif key == "iyr":
            year = int(passport[key])
            if((year < 2010) or (year > 2020)):
                return False
        elif key == "eyr":
            year = int(passport[key])
            if((year < 2020) or (year > 2030)):
                return False
        elif key == "hgt":
            height = passport[key]
            if("cm" not in height) and ("in" not in height):
                return False
            if "cm" in height:
                cm_height = int(height.strip("cm"))
                if((cm_height < 150) or (cm_height > 193)):
                    return False
            if "in" in height:
                in_height = int(height.strip("in"))
                if((in_height < 59) or (in_height > 76)):
                    return False
        elif key == "hcl":
            if not re.match("#[a-f0-9]{6}$",passport[key]):
                return False
        elif key == "ecl":
            if passport[key] not in ["amb", "blu", "brn", "gry", "grn", "hzl", "oth"]:
                return False
        elif key == "pid":
            if not re.match("[0-9]{9}$", passport[key]):
                return False
    return True

def part_two(data, req_fields):
    counter = 0
    for passport in data:
        if validate_passport(passport, req_fields):
            counter += 1
    return counter

nice_data = make_dict(raw_data2)
print(part_one(nice_data, check_fields))
print(part_two(nice_data, check_fields))